### DISCUSSION TITLE
Fennel integration

### DISCUSSION THREAD

**Discussion by caligian** [NTBBloodbath/doom-nvim#280] [Ideas]
I am revamping my configuration with fennel.

I feel that it is absolutely fantastic and clean to configure in fennel. I think that we should implement some core functionality via fennel. The usage of macros is a blessing in many cases.

I shall post a utils.fnl which is a rewrite of utils.lua of mine soon.

**Comment by molleweide** [comment]
I like!! I also want to learn fennel so having small chunks of fennel would be great for learning

**Comment by NTBBloodbath** [comment]
Hi caligian!

Tbh for months I have a whim with an idea that for many may be abominable, that idea is rewriting doom in fennel haha. After a small amount of time thinking about this I found some possible problems on it.

1. Lua is well known in the neovim community nowadays (at least by people contributing to its ecosystem) so is a better standard than fennel to find contributors.
2. Fennel gets transpiled into Lua at the end of the day and needs to be transpiled on each change to take effect, I'm ok with this but you know, this behavior can be a bit annoying for some people.

Writing only some parts of the code? I feel like it's fine if you're going to maintain them by yourself (I could help too, fennel isn't that hard) due to first problem (lack of fennel devs in neovim community).

I'm not completely rejecting this idea, I currently find it good but I think this needs to be properly discussed because it introduces a new language to our stack and like everything else it introduces a learning curve for our current _and future_ contributors.

Also, is there a specific list of functionalities you want to rewrite in Fennel? Just being curious haha.

Cheers!

**Reply by caligian** [reply]
I think you are right. I have dabbled in elisp and clojure and hence found fennel easy to pick up but that may not be the case for many people.

However...
Fennel transcompiles to lua with merely 2 lines code. So it is easy for others to use fennel code in their lua source code given that at least some documentation is available on what the functions do. 

I am pro-fennel because of a few things but these are just my opinions:

1. Functional paradigm is noticeably cleaner and encourages declarative style which makes code comprehension rather easy as the logic is laid out pretty nicely. Eg. map and reduce are much much easier to grok than a long iterative block. I managed to replace a lot of repetitive code in fennel with luafun (excellent functional library) and sometimes with simple recursion. While the same can be done in lua, fennel code is more elegant. 
2. Fennel has optional strict parameter checking where only a specific number of parameters are accepted and nils throw an error. You have to suffix a parameter with ? in order to skip it and you can do that to ANY parameter (even the first). That is a great feature because it makes the code cleaner.
3. Named lambdas !?!
4. 100% compat with lua libs  
5. Macros!
6. A nice pattern matching construct - 'match' which is like perl6's smart operator and ruby's switch-case

The drawbacks:
1. Functional paradigm is difficult to implement in practice even for experienced coders 
2. It requires a completely different style of thinking which is a hindrance when you want to port code because you have to completely rewrite many things. I am experiencing the pain first-hand :)
3. Macros are easy to hate. 
4. Configuration is preferred to be rather quickly written than painfully laid out like a typical plugin. 

So, we can do one thing. Fennel coders can augment the current config by creating utility functions for the developers and the users. My new implementation of hook and kbd is ready without any change in the way they handle parameters. 

I am also planning to write some more functions which are similar to doom's after!, native indentation code and a basic REPL. This time the REPL will actually handle the handle and show it in a temp buffer or try to implement something like conjure (opening a popup window for output)

I think this make us more productive?

**Reply by connorgmeehan** [reply]
Hey just adding my part.  I'm personally against partially adding fennel to the project because I think it'd introduce complexity and add to the development time.  The extra language features are appealing to me and I'm interested in learning a lisp family language but I think adding an abstraction to part of the code base is going to make it harder to develop.  Also I don't love the idea of excluding contributors (current or future) who aren't confident in fennel.

I agree that functional code blocks are easier to read, there are some functional programming utilities in the `vim` object that we could make more use of such as `vim.tbl_filter` and `vim.tbl_map` + adding our own reduce methods.  Alternatively, Luafun looks awesome and having these functional utilities could be really helpful in places.  I think it's worth assessing if and where it would help after the new module architecture is merged.
 
I've also heard the same thing about the functional paradigm being difficult to master, I think I'd be more open to it if someone who has more experience could drive this implementation.  At the end of the day I just want a stable up to date text editor and I'm not sure that this will help.

I guess I am pretty skeptical but interested in hearing more so I have some questions.
How does a project that is partially written in fennel work?
Which parts of doom-nvim would suit being written in fennel?

**Reply by caligian** [reply]
I rewrote everything in fennel. Do you want to take a peek?

Fennel rather reduced development time for me. Only complexity was to fiddle with package.path and writing a recompiler for fnl configs

**Reply by connorgmeehan** [reply]
Yeah I'd be keen to have a look? Are you pushing to a repo on your profile?

**Reply by caligian** [reply]
[doom-evim](https://github.com/caligian/doom-evim)
Here you go.

**Comment by caligian** [comment]
In retrospect, it seems like a stupid idea. I trashed the fennel rewrite. Errors are difficult to debug because it points to the line numbers in the transpiled code :(
